adding sklearn-style PAC estimator

choldgraf opened the pull request:
Is this what you had in mind @kingjr? LMKWYT and I can make edits. If it looks good I can add tests.

choldgraf commented:
I should add - right now the PAC function doesn't support Epochs-shaped data. It seems like it'd be hard to support this with the sklearn-style API as well, no? Since sklearn transformers expect 2D data as input and epochs are 3D?

kingjr commented:
yes, thanks. 

Let's wait though, I'm wondering whether we should to manually create a transformer for each transformation, or whether we should just apply a Mixin.

kingjr commented:
> 

I should add - right now the PAC function doesn't support Epochs-shaped data. It seems like it'd be hard to support this with the sklearn-style API as well, no? Since sklearn transformers expect 2D data as input and epochs are 3D?


ouch, I hadn't realized that. Why cannot it handle 3D arrays?

choldgraf commented:
> ouch, I hadn't realized that. Why cannot it handle 3D arrays?

it theoretically could w/o a ton of extra overhead but the internals of how the function works would need to change. The main reasons were:

1. The PAC PR was already pretty beefy and that added another layer of complication
2. It added extra challenges w/r/t edge artifacts and such (we'd need to add checks that the windows weren't too small for example)

So I decided to make you use `Raw` data, and gave the option to pass an `events` parameter, that way all the filtering is done beforehand, and then the data is chunked into trials after the fact.

I just modified the code a bit to (I think) allow for epochs-style input (though the shapes would be a little off since it expects (n_epochs, n_times, n_channels), though I could change that if we don't need to be strict about passing sklearn-style inputs (samples x channels)) WDYT?

kingjr commented:
For sklearn-like transformers, I think we'll speak to a more advance userbase who don't want to be limited by edge warnings and such, as they'll just use this step to see whether it can extract additional information.

I think it'd make more sense to have n_epochs, n_chans, n_times to PAC.fit(X); what is done internally doesn't matter that much.

choldgraf commented:
ok cool - so w/ the current setup you could do:

```python
ixs = [(1, 2), (2, 3)]
f_ph = [(4, 8)]
f_am = [(60, 80)]
sfreq = 1000.
tmin, tmax = 0, .5
times = epochs.times
pac = PAC(f_ph, f_am, ixs, sfreq, times=times, tmin=tmin, tmax=tmax)
pac.fit(epochs.get_data())
pac_values = pac.transform()
# pac_values is shape (n_epochs, n_freq_pairs * n_time_wins * n_chan_pairs)

kingjr commented:
perfect

choldgraf commented:
ok, now I just need to write a test and make sure that my above statement is actually correct 😆

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+                (n_epochs, n_channels, n_times)
+            The input data.
+        """
+        if self.events is not None:
```
put this check in a private method, to call it at transform() too

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        self.ixs = ixs
+        self.sfreq = sfreq
+        self.pac_func = pac_func
+        self.events = events
```
why?

choldgraf replied:
I thought the only parameters we wanted to allow to the `fit` or `transform` methods were X/y...

kingjr replied:
Yes, it's a sklearn corner case ATM. But putting it in init isn't the answer, because transformer should be applicable to different dataset without reinit.

I think we should not implement the raw version for now; we'll wait until sklearn solves the possibility to pass more than X and y in their pipeline.

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        X : array, shape (n_channels, n_times) |
+                (n_epochs, n_channels, n_times)
+            The input data.
+        """
```
and y : None | array for sklearn compatibility purposes.

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+            if X.ndim == 3:
+                raise ValueError('If Events is not None, X must be 2D')
+            if self.events[:, 0].max() > X.shape[0]:
+                raise ValueError('Event index exceeds dimensions of X.')
```
what does this do?

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+            # Add an extra first dimension for the loop
+            X = [X]
+
+        pac = []
```
can you init a numpy array of the right size? or this is difficult to guess?

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+                self.n_cycles_ph, self.n_cycles_am, self.scale_amp_func)
+            # Reshape PAC so it's 2D and first dimension is Epochs
+            pac.append(i_pac.reshape(pac.shape[0], -1))
+        self.pac = np.asarray(pac)
```
no don't store the values. nothing happens at fitting, it's just a transform

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        return self.pac
+
+    def fit_transform(self, X, y=None):
+        self.fit(X)
```
move the main computation here and return the result.

the sklearn object only stores fitted coefficients, not transformed values.

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+            pac.append(i_pac.reshape(pac.shape[0], -1))
+        self.pac = np.asarray(pac)
+
+    def transform(X=None, y=None, self):
```
If we want to have raws with events, events must be passed here. Although adding ad hoc param is sklearn compatible it is currently not pipelineable, so I suggest to keep this for later.

In other words, let's not support raw data for now.

choldgraf replied:
so then basically `fit` and `transform` will do nothing, and `fit_transform` will do all the work?

kingjr replied:
Close: fit does nothing, transform does all the work, fit_transform calls transform

choldgraf commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +121,129 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        self.n_cycles_ph = n_cycles_ph
+        self.n_cycles_am = n_cycles_am
+        self.scale_amp_func = scale_amp_func
+        self.n_jobs = n_jobs
```
ya, I'll beef up the robustness of the class if we're happy w/ the API

choldgraf replied:
isn't that going to duplicate all the checks that are inside `_phase_amplitude_coupling` though?

kingjr replied:
To avoid redundancy, you can externalize the check of `_phase_amplitude_coupling` and use them here, but the params should be checked at init IMO.

choldgraf replied:
I guess I can throw that into a private function and call that in both places instead. It'll get called twice but that's probably not a big deal

choldgraf replied:
oops, looks like our comments crossed - OK I can do this

kingjr replied:
yes. Only 37 seconds late. :]

choldgraf commented:
thoughts on latest version? I think most comments are addressed.

choldgraf commented:
still need to add tests

kingjr commented on examples/connectivity/plot_pac_est.py:
```diff
@@ -0,0 +1,131 @@
+          interpolation='nearest')
+ax.set_ylabel('Epoch')
+ax.set_xlabel('PAC feature')
+plt.show()
```
Mmmh, a bit too redundant with the pac example no?

I was thinking of an advanced pipeline, but we need to wait for the other transformers to be finished:
```
make_feature_union(
     make_pipeline(Filterer(theta[0], theta[1], sfreq=sfreq),
                              UnsupervisedSpatialFilter(ICA(10))),
     make_pipeline(Filterer(gamma[0], gamma[1], sfreq=sfreq), 
                              CSP(10),
                              Rectifier(),
                              Filterer(0, 20, sfreq=sfreq),
)
```

choldgraf replied:
agreed it's redundant - this is sort of a minimum-viable product version of an example, and I copy/pasted most of the code from the other one. It's just meant to show the API, but I agree it'd be much better if we could show it in a proper chain. I just don't have any data that would show a good classification performance for PAC values.

kingjr replied:
ok, so perhaps let's keep it aside for now, or include the transformer inside the already existing example

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        self.n_jobs = n_jobs
+
+    def fit(self, X=None, y=None):
+        """
```
Does nothing, only used for scikit-learn compatibility.

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        Parameters
+        ----------
+        X : array, shape (n_channels, n_times) |
+                (n_epochs, n_channels, n_times)
```
only latter shape

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+            The input data.
+        y : None | array
+        """
+        n_ep = X.shape[0]
```
check ndim == 3 here

kingjr replied:
uses n_epochs

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        n_t_wins = len(self.tmin)
+        n_ch_pairs = len(self.ixs)
+
+        pac = np.zeros([X.shape[0], n_f_pairs * n_t_wins * n_ch_pairs])
```
n_epochs

kingjr replied:
use np.empty, so that if you have overlapping jobs they recruits less RAM

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        y : None | array
+        """
+        n_ep = X.shape[0]
+        n_f_pairs = len(self.f_amp) * len(self.f_phase)
```
n_freq_pairs is a bit more readable IMO

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        """
+        n_ep = X.shape[0]
+        n_f_pairs = len(self.f_amp) * len(self.f_phase)
+        n_t_wins = len(self.tmin)
```
n_time_wins

kingjr replied:
add comment on the side to have explicit naming

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+                ep_X, self.sfreq, self.f_phase, self.f_amp, self.ixs,
+                self.pac_func, None, self.tmin, self.tmax, self.times,
+                self.n_cycles_ph, self.n_cycles_am, self.scale_amp_func)[0]
+            # Reshape PAC so it's 2D and first dimension is Epochs
```
no, keep it n dimensional, sklearn only needs the first dimension to be n_samples, the rest can be pipe as ndimensional arrays

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        X : array, shape (n_channels, n_times) |
+                (n_epochs, n_channels, n_times)
+            The input data.
+        y : None | array
```
Not used, for scikit-learn compatibility purposes.

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        Parameters
+        ----------
+        X : array, shape (n_channels, n_times) |
+                (n_epochs, n_channels, n_times)
```
only latter shape no?

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
 def _phase_amplitude_coupling(data, sfreq, f_phase, f_amp, ixs,
                               pac_func='ozkurt', events=None,
-                              tmin=None, tmax=None, n_cycles_ph=3,
+                              tmin=None, tmax=None, times=None, n_cycles_ph=3,
```
are tmin and tmax necessary for epochs?

I think it's easier just to add a step in the pipeline e.g. `make_pipeline(TimeSelector(), PAC())` to reduce number of params

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        self.f_phase = f_phase
+        self.f_amp = f_amp
+        self.tmin = tmin
+        self.tmax = tmax
```
are tmin and tmax necessary for epochs?

I think it's easier just to add a step in the pipeline e.g. make_pipeline(TimeSelector(), PAC()) to reduce number of params

kingjr replied:
Except if this avoids edges and not my `TimeSelector` idea

choldgraf replied:
The tmin / tmax parameters let you supply multiple timepoints for PAC calculation, and yeah they also avoid the problem of edge artifacts. E.g., you'd want to supply Epochs w/ times.min() = -1, and times.max() = 1, and then supply tmin/tmax of `[0, .1]` or something. Otherwise we might get edge artifacts...also the pacpy filtering code is a little wonky and has a tendency to return exceptions because of window length issues.

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        self.f_amp = f_amp
+        self.tmin = tmin
+        self.tmax = tmax
+        self.times = times
```
see above comment

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        self.tmin = tmin
+        self.tmax = tmax
+        self.times = times
+        self.ixs = ixs
```
what is this for?

choldgraf replied:
it's the channel indices for pairs of channels in the PAC estimation. I'll rename to `ch_ixs`

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+    n_jobs : int
+        Number of jobs to run in parallel. Defaults to 1.
+    """
+    def __init__(self, f_phase, f_amp, ixs, sfreq, times,
```
default sfreq = 1

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -121,17 +122,136 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        ----------
+        X : None | array
+        y : None | array
+        """
```
Return self, see sklearn convention

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -332,16 +428,78 @@ def _phase_amplitude_coupling(data, sfreq, f_phase, f_amp, ixs,
+    err = 'Must supply one of events or times (not both)'
+    if all(ii is not None for ii in [events, times]):
+        raise ValueError(err)
+    if all(ii is None for ii in [events, times]):
```
There must be a more pythonic way doing this. Maybe `None == events == times`

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -332,16 +428,78 @@ def _phase_amplitude_coupling(data, sfreq, f_phase, f_amp, ixs,
+def _check_pac_times(events, times, tmin, tmax):
+    # Times and time windows
+    err = 'Must supply one of events or times (not both)'
+    if all(ii is not None for ii in [events, times]):
```
None not in (events, times)?

kingjr replied:
nope

choldgraf replied:
what about `(events is times is None) or (events is not times is not None)`

choldgraf replied:
actually that doesn't work, but I think `if (events is times is None) or (events is not None and times is not None)` does

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -332,16 +428,78 @@ def _phase_amplitude_coupling(data, sfreq, f_phase, f_amp, ixs,
+            raise ValueError('events have incorrect number of columns')
+        if any([ii is None for ii in [tmin, tmax]]):
+            raise ValueError('If events is given, tmin/tmax must be given.')
+    if times is not None:
```
times = np.asarray(times) to accept lists

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -332,16 +428,78 @@ def _phase_amplitude_coupling(data, sfreq, f_phase, f_amp, ixs,
+
+
+def _check_pac_data(ixs, f_phase, f_amp, n_cycles_ph, n_cycles_am, pac_func):
+    # PAC channel indices
```
a bit of docstring would help, I'm not sure what ixs is

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -332,16 +428,78 @@ def _phase_amplitude_coupling(data, sfreq, f_phase, f_amp, ixs,
+    if f_phase.shape[-1] != 2 or f_amp.shape[-1] != 2:
+        raise ValueError('Frequencies must be specified w/ a low/hi tuple')
+
+    for icyc, i_f in zip([n_cycles_ph, n_cycles_am], [f_phase, f_amp]):
```
for icyc, i_f in ((n_cycles_ph, f_phase), (n_cycles_am, f_amp)):

kingjr commented:
In the right direction. I think we still need to discuss the example though

choldgraf commented:
cool - I think I addressed comments above. I moved the example back to be at the end of one of the previous ones. LMKWYT. I need to work on some other stuff today but if I have time I'll get back to this later this evening or tomorrow.

kingjr commented:
Ok, no time left for me before Friday. Re-ping me if you see that I forget about it

choldgraf commented:
Sounds good - it looks like the error is due to something other than my addition. LMK if you have other thoughts on this and then I can take a look again early next week.

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -114,24 +119,164 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+    ch_ixs : array-like, shape (n_ch_pairs x 2)
+        The indices for low/high frequency channels. PAC will be estimated
+        between n_ch_pairs of channels. Indices correspond to rows of `data`.
+    times : array, shape (n_times,) | None
```
shouldn't this be default to range(len(signals))?

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -114,24 +119,164 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+    n_cycles_ph : float, int | array of floats, shape (n_bands_phase,)
+        The number of cycles to be included in the window for each band-pass
+        filter for phase. Defaults to 3.
+    n_cycles_am : float, int | array of floats, shape (n_bands_amp,)
```
n_cycles_amp for consistency

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -114,24 +119,164 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
-def _phase_amplitude_coupling(data, sfreq, f_phase, f_amp, ixs,
+    Attributes
+    ----------
+    freq_pairs_ : array, shape (n_freq_pairs, 2, 2)
```
can you clarify what the 2 and 2 are?

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -114,24 +119,164 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        """
+        if X.ndim != 3:
+            raise ValueError('X must be shape (n_epochs, n_channels, n_times)')
+        n_epochs = X.shape[0]
```
len(X)

kingjr commented on mne_sandbox/connectivity/cfc.py:
```diff
@@ -114,24 +119,164 @@ def phase_amplitude_coupling(inst, f_phase, f_amp, ixs, pac_func='ozkurt',
+        return pac
+
+    def fit_transform(self, X, y=None, times=None):
+        """Not used. For sklearn compatibility purposes only.
```
no, it does transform the data, copy transform docstring

kingjr commented:
Minor comments, LGTM.

choldgraf commented:
Cool - I wanna take a pass through and clean up some code a little bit. I
get the feeling that I'm doing things unnecessarily complicated in there.
But otherwise I'll take a look at your edits and then perhaps we can merge.
Between this and the encoding model PR I've definitely set a personal
record for length of time to finish pull requests :P

On Thu, Aug 11, 2016 at 6:59 AM Jean-Rémi KING <notifications@github.com>
wrote:

> Minor comments, LGTM.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-sandbox/pull/16#issuecomment-239169073>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwSHXxhs2dJtGNgL9mU0lksQvrCgfB_ks5qeyrUgaJpZM4JVoxT>
> .
>

choldgraf commented:
Addressed the comments above and added a few tests for the `PAC` estimator as well. I think the code could probably still be improved but maybe it's better to just merge it and see how/if people use it

kingjr commented:
LGTM

Le 12 août 2016 1:43 AM, "Coveralls" <notifications@github.com> a écrit :

> [image: Coverage Status] <https://coveralls.io/builds/7417296>
>
> Coverage increased (+0.8%) to 92.671% when pulling *75fdaf4
> <https://github.com/mne-tools/mne-sandbox/commit/75fdaf48b3f22456c4dd26fc92d0a47bf7480bd4>
> on choldgraf:cfc_sklearn* into *e1310b4
> <https://github.com/mne-tools/mne-sandbox/commit/e1310b418a0c724336748bad6cc5ac6f0d2c0e03>
> on mne-tools:master*.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-sandbox/pull/16#issuecomment-239365575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEp7DOWwpTiIWiJ7AQxYEz0mjy8IWjkzks5qfAgFgaJpZM4JVoxT>
> .
>

choldgraf commented:
Let me know if there's anything more to do on my end here...

kingjr commented:
@agramfort, do you have time to look at this/opinion?

choldgraf commented:
(though if you have very limited time, probably better for somebody to look
at the encoding model PR in mne-python before looking at this one, as it's
a little more immediately relevant I think)

-- 

On Mon, Aug 15, 2016 at 10:37 AM Jean-Rémi KING <notifications@github.com>
wrote:

> @agramfort <https://github.com/agramfort>, do you have time to look at
> this/opinion?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-sandbox/pull/16#issuecomment-239870799>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwSHced4SJE2tJIZWnS5zDOzOvuyygaks5qgKPwgaJpZM4JVoxT>
> .
>

choldgraf commented:
shall I close this in lieu of the other package @agramfort ?

agramfort commented:
if this new package does a better jobs yes. Please test.

choldgraf commented:
I'm gonna close this since @agramfort mentioned his student has a better implementation in the works.

kingjr commented:
Ok

On Thursday, 27 April 2017, Chris Holdgraf <notifications@github.com> wrote:

> I'm gonna close this since @agramfort <https://github.com/agramfort>
> mentioned his student has a better implementation in the works.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-sandbox/pull/16#issuecomment-297778062>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEp7DIguDCEBqwOjUVcDGaMyX1kDVXfgks5r0MtBgaJpZM4JVoxT>
> .
>
